#  T-tests and P-values

In statistics, t-test is used to test if two data samples have a significant difference between their means. There are two types of t-test:

* **Student's t-test** (a.k.a. independent or uncorrelated t-test). This type of t-test is to compare the samples of **two independent populations** (e.g. test scores of students in two different classes). `scipy` provides the [`ttest_ind`](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ttest_ind.html) method to conduct student's t-test.

* **Paired t-test** (a.k.a. dependent or correlated t-test). This type of t-test is to compare the samples of **the same population** (e.g. scores of different tests of students in the same class). `scipy` provides the [`ttest_re`](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ttest_rel.html) method to conduct paired t-test.

Both types of t-tests return a number which is called the **p-value**. If p-value is below 0.05, we can confidently declare the null-hypothesis is rejected and the difference is significant. If p-value is between 0.05 and 0.1, we may also declare the null-hypothesis is rejected but we are not highly confident. If p-value is above 0.1 we do not reject the null-hypothesis.

Read more about the t-test in [this article](http://b.link/test50) and [this Quora](http://b.link/unpaired97). Make sure you understand when to use which type of t-test. 

![logo_ironhack_blue 7](https://user-images.githubusercontent.com/23629340/40541063-a07a0a8a-601a-11e8-91b5-2f13e4e6b441.png)

# Lab | Inferential statistics - T-test & P-value

### Instructions

**1. *One tailed t-test* - In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file `files_for_lab/machine.txt`.
   Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?**

In [33]:
# Import libraries
import pandas as pd
import numpy as np
import scipy.stats as st

In [34]:
##############Hypotheses############

#H0 - The new machine's packing time is the same or more, on average
#H1 - The new machine's packing time is faster, on average

In [35]:
# import data
data = pd.read_csv('machine2.txt', sep = ' ')
data

New_machine  Old_machine
0         42.1         42.7
1         41.0         43.6
2         41.3         43.8
3         41.8         43.3
4         42.4         42.5
5         42.8         43.5
6         43.2         43.1
7         42.3         41.7
8         41.8         44.0
9         42.7         44.1

In [36]:
print("new machine mean is: ",round(data['New_machine'].mean(),2))
print("old machine mean is: ",round(data['Old_machine'].mean(),2))

new machine mean is:  42.14
old machine mean is:  43.23


In [37]:
print('p value (single tailed): ',st.ttest_1samp(data['New_machine'],43.23)[1]/2)

p value (single tailed):  0.0003483188038379669


The 0 Hypothesis is a lot less than 0.05, H0 is rejected.


**2. *Matched Pairs Test* - In this challenge we will compare dependent samples of data describing our Pokemon (file `files_for_lab/pokemon.csv`). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.**

In [38]:
##############Hypotheses############

#H0 - Defense and attack scores are the same
#H1 - Defense and attack scores are different

In [39]:
pokemon = pd.read_csv('pokemon.csv')
pokemon = pokemon.drop(columns='#')
pokemon

Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0                Bulbasaur    Grass  Poison    318  45      49       49   
1                  Ivysaur    Grass  Poison    405  60      62       63   
2                 Venusaur    Grass  Poison    525  80      82       83   
3    VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4               Charmander     Fire     NaN    309  39      52       43   
..                     ...      ...     ...    ...  ..     ...      ...   
795                Diancie     Rock   Fairy    600  50     100      150   
796    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [40]:
st.ttest_rel(pokemon['Attack'], pokemon['Defense'])

Ttest_relResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)

In [41]:
#E-05, the H1 is alive! We reject H0 again, as the pvalue < 0.05 

In [42]:
# Import libraries
import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns

#### Import dataset

In this challenge we will work on the Pokemon dataset you have already used. The goal is to test whether different groups of pokemon (e.g. Legendary vs Normal, Generation 1 vs 2, single-type vs dual-type) have different stats (e.g. HP, Attack, Defense, etc.). Use Ironhack's database to load the data (db: pokemon, table: pokemon_stats). 

In [43]:
pokemon = pd.read_csv('pokemon.csv')
pokemon = pokemon.drop(columns='#')
pokemon

Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0                Bulbasaur    Grass  Poison    318  45      49       49   
1                  Ivysaur    Grass  Poison    405  60      62       63   
2                 Venusaur    Grass  Poison    525  80      82       83   
3    VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4               Charmander     Fire     NaN    309  39      52       43   
..                     ...      ...     ...    ...  ..     ...      ...   
795                Diancie     Rock   Fairy    600  50     100      150   
796    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

#### First we want to define a function with which we can test the means of a feature set of two samples. 

In the next cell you'll see the annotations of the Python function that explains what this function does and its arguments and returned value. This type of annotation is called **docstring** which is a convention used among Python developers. The docstring convention allows developers to write consistent tech documentations for their codes so that others can read. It also allows some websites to automatically parse the docstrings and display user-friendly documentations.

Follow the specifications of the docstring and complete the function.

In [44]:
legendary = pokemon.loc[pokemon['Legendary'] == True]
normal = pokemon.loc[pokemon['Legendary'] == False]

In [45]:
features=['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Total']

In [46]:
s_1 = legendary[features]
s_2 = legendary[features]

In [47]:
def t_test_features(s_1, s_2, features=['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Total']):
    """Test means of a feature set of two samples
   
    Args:
        s1 (dataframe): sample 1
        s2 (dataframe): sample 2
        features (list): an array of features to test
    Returns:
        dict: a dictionary of t-test scores for each feature where the feature name is the key and the p-value is the value
    """

    # Your code here
    results = {}
    for feature in features:
        results[feature] = st.ttest_ind(s_1[feature], s_2[feature], equal_var = False).pvalue
    return results


#### Using the `t_test_features` function, conduct t-test for Lengendary vs non-Legendary pokemons.

*Hint: your output should look like below:*

```
{'HP': 1.0026911708035284e-13,
 'Attack': 2.520372449236646e-16,
 'Defense': 4.8269984949193316e-11,
 'Sp. Atk': 1.5514614112239812e-21,
 'Sp. Def': 2.2949327864052826e-15,
 'Speed': 1.049016311882451e-18,
 'Total': 9.357954335957446e-47}
 ```

In [48]:
t_test_features(s_1, s_2)

{'HP': 1.0,
 'Attack': 1.0,
 'Defense': 1.0,
 'Sp. Atk': 1.0,
 'Sp. Def': 1.0,
 'Speed': 1.0,
 'Total': 1.0}

#### From the test results above, what conclusion can you make? Do Legendary and non-Legendary pokemons have significantly different stats on each feature?

No, all the p-values are << (much lower) than 0.05 so we can conclude the variances of the 2 samples are really small

They dont have significantly different stats between LEGENDARY and NORMAL

#### Next, conduct t-test for Generation 1 and Generation 2 pokemons.

In [49]:
# Your code here
type_1 = pokemon.loc[pokemon['Generation'] == 1]

type_2 = pokemon.loc[pokemon['Generation'] == 2]

t_test_features(type_1, type_2)

{'HP': 0.14551697834219623,
 'Attack': 0.24721958967217725,
 'Defense': 0.5677711011725426,
 'Sp. Atk': 0.12332165977104394,
 'Sp. Def': 0.18829872292645752,
 'Speed': 0.00239265937312135,
 'Total': 0.5631377907941676}

#### What conclusions can you make?

No, all the p-values are << (much lower) than 0.05 so we can conclude the variances of the 2 samples are really small

They dont have significantly different stats between GENERATION1 and GENERAIION2 

#### Compare pokemons who have single type vs those having two types.

In [50]:
# Your code here
single_type = pokemon.loc[pokemon['Type 2'].isnull()]

dual_type = pokemon.loc[pokemon['Type 2'].notnull()]

t_test_features(single_type, dual_type)

{'HP': 0.11314389855379421,
 'Attack': 0.00014932578145948305,
 'Defense': 2.7978540411514693e-08,
 'Sp. Atk': 0.00013876216585667842,
 'Sp. Def': 0.00010730610934512779,
 'Speed': 0.02421703281819094,
 'Total': 1.1157056505229961e-07}

#### What conclusions can you make?

No, all the p-values are << (much lower) than 0.05 so we can conclude the variances of the 2 samples are really small

They dont have significantly different stats between GENERATION1 and GENERATION2 

#### Now, we want to compare whether there are significant differences of `Attack` vs `Defense`  and  `Sp. Atk` vs `Sp. Def` of all pokemons. Please write your code below.

*Hint: are you comparing different populations or the same population?*

#### What conclusions can you make?

In [51]:
# H0: no relevant difference between Attack and Defense
# H1: relevant difference 
print(st.ttest_rel(pokemon.Attack, pokemon.Defense))

Ttest_relResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)


**p-value=1.71 > 0.05 => We do not reject H0, NO RELEVANT DIFFERENCE BETWEEN Attack and Defense**

In [52]:
# H0: no relevant difference between Sp. Atk and Sp. Def
# H1: relevant difference 

print(st.ttest_rel(pokemon['Sp. Atk'], pokemon['Sp. Def']))

Ttest_relResult(statistic=0.853986188453353, pvalue=0.3933685997548122)


**p-value=0.3933 > 0.05 => We reject H0, THERE IS A RELEVANT DIFFERENCE BETWEEN Sp. Atk and Sp. Def**

#  ANOVA

In statistics, **Analysis of Variance (ANOVA)** is also used to analyze the differences among group means. The difference between t-test and ANOVA is the former is ued to compare two groups whereas the latter is used to compare three or more groups. [Read more about the difference between t-test and ANOVA](http://b.link/anova24).

From the ANOVA test, you receive two numbers. The first number is called the **F-value** which indicates whether your null-hypothesis can be rejected. The critical F-value that rejects the null-hypothesis varies according to the number of total subjects and the number of subject groups in your experiment. In [this table](http://b.link/eda14) you can find the critical values of the F distribution. **If you are confused by the massive F-distribution table, don't worry. Skip F-value for now and study it at a later time. In this challenge you only need to look at the p-value.**

The p-value is another number yielded by ANOVA which already takes the number of total subjects and the number of experiment groups into consideration. **Typically if your p-value is less than 0.05, you can declare the null-hypothesis is rejected.**

In this challenge, we want to understand whether there are significant differences among various types of pokemons' `Total` value, i.e. Grass vs Poison vs Fire vs Dragon... There are many types of pokemons which makes it a perfect use case for ANOVA. Use Ironhack's database to load the pokemon data (db: pokemon, table: pokemon_stats). 

# Lab | Inferential statistics - ANOVA

Note: The following lab is divided in 2 sections.

## Part 1

In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on:
    - Null hypothesis
    - Alternate hypothesis
    - Level of significance
    - Test statistic
    - P-value
    - F table

### Context

Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data `anova_lab_data.xlsx` file in the `files_for_lab` folder  

Data was collected randomly and provided to you in the table as shown: [link to the image - Data](https://education-team-2020.s3-eu-west-1.amazonaws.com/data-analytics/7.05/7.05-lab_data.png)


In [53]:
# Import libraries
import pandas as pd
import numpy as np
import scipy.stats as st

#### - State the null hypothesis

In [54]:
# H0 - There is no difference in the mean etching rate

#### - State the alternate hypothesis

In [55]:
# H1 - There is a difference in the mean etching rate

#### - What is the significance level

In [56]:
# 0.05 - It's not an urgent matter (which would require a lower level of significance) 

#### - What are the degrees of freedom of model, error terms, and total DoF

## Part 2

- In this section, use the Python to conduct ANOVA.
- What conclusions can you draw from the experiment and why?

In [57]:
micro = pd.read_excel('anova_lab_data.xlsx')
micro

Power   Etching Rate
0   160 W          5.43
1   180 W          6.24
2   200 W          8.79
3   160 W          5.71
4   180 W          6.71
5   200 W          9.20
6   160 W          6.22
7   180 W          5.98
8   200 W          7.90
9   160 W          6.01
10  180 W          5.66
11  200 W          8.15
12  160 W          5.59
13  180 W          6.60
14  200 W          7.55

- Test statistic

In [58]:
micro['power_rate'] = micro.groupby('Power ').cumcount() ##is the new index 
micro_pivot = micro.pivot(index='power_rate', columns='Power ', values='Etching Rate')
micro_pivot.columns = ['Power_'+str(x) for x in micro_pivot.columns.values]
micro_pivot.head()

Power_160 W  Power_180 W  Power_200 W
power_rate                                       
0                  5.43         6.24         8.79
1                  5.71         6.71         9.20
2                  6.22         5.98         7.90
3                  6.01         5.66         8.15
4                  5.59         6.60         7.55

- P-value 

In [59]:
st.f_oneway(micro_pivot['Power_160 W'],micro_pivot['Power_180 W'],micro_pivot['Power_200 W'])

F_onewayResult(statistic=36.87895470100505, pvalue=7.506584272358903e-06)

Here the pvalue is essentially 0, so we must reject our H0 hypothesis. 